File: java_solutions/494.target-sum.java
/*
 * @lc app=leetcode id=494 lang=java
 *
 * [494] Target Sum
 */

record Solution() {
    public int findTargetSumWays(int[] nums, int target) {
        return java.util.stream.IntStream.of(nums).sum() < Math.abs(target) || (java.util.stream.IntStream.of(nums).sum() + target) % 2 != 0 ? 0 : java.util.Arrays.stream(nums).reduce(new int[(java.util.stream.IntStream.of(nums).sum() + target) / 2 + 1], (dp, n) -> java.util.stream.Stream.of(dp).peek(arr -> arr[0] = arr[0] == 0 ? 1 : arr[0]).peek(arr -> java.util.stream.IntStream.iterate(arr.length - 1, i -> i >= n, i -> i - 1).forEach(i -> arr[i] += arr[i - n])).findFirst().get(), (a, b) -> a)[(java.util.stream.IntStream.of(nums).sum() + target) / 2];
    }
}

File: kotlin_solutions/494.target-sum.kt
/*
 * @lc app=leetcode id=494 lang=kotlin
 *
 * [494] Target Sum
 */

class Solution { fun findTargetSumWays(nums: IntArray, target: Int): Int = nums.fold(mapOf(0 to 1)) { dp, num -> mutableMapOf<Int, Int>().apply { dp.forEach { (sum, cnt) -> put(sum + num, getOrDefault(sum + num, 0) + cnt) .also{put(sum - num, getOrDefault(sum - num, 0) + cnt)} } } }.getOrDefault(target, 0) }